### DISCUSSION TITLE
access token with permissions that differ by repo

### DISCUSSION THREAD

**Discussion by petar** [community/community#66930] [Actions] [answered]
### Select Topic Area

Question

### Body

I am trying to create a (fine-grain) access token which has the following permissions:
- permission to create new repos in a given org
- permission to read issues from a specific existing repo
Is this possible with the current GitHub fine grain tokens?

Alternatively, is it possible to create an access token that:
- can write to repo1
- can read from repo2
I didn't find a way to do this either.

Thank you
Petar

**Comment by ryanmio** [comment, marked as answer]
Tokens' scopes are org-wide or user-wide, depending on how you create them. You may need to use a combination of GitHub Actions and GitHub API to manually manage permissions. It's not ideal, but that's the current limitation

**Comment by elendil7** [comment]
## Abstract

I will attempt to answer the 2nd part of your question: 

`Alternatively, is it possible to create an access token that:
    1. can write to repo1
    2. can read from repo2`

For this setup, I'd recommend creating per-repository **deploy keys**, which coincidentally have the exact access control mechanisms you require: *READ-only*, or *READ-WRITE*.

These tokens can be created at the following link:
`https://github.com/<organisationnameORusername>/<reponame>/settings/keys/new`

You must first generate the private-public keys locally using something like **ssh-keygen**. Note that you [cannot use a single key across all repositories](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/managing-deploy-keys#using-multiple-repositories-on-one-server), meaning in the case of 2 repos, you can generate 2 keys locally (one per repo).

Then you must add the generated **public** key to the "key" field of the form. Give the deploy key a fitting title, and also check the "allow write access" if you want your key to have both read and write access to the repo. Otherwise, if you leave the box unchecked, the key will only have read access.

If you're on windows / linux, you could try the following approach, which involves defining a singular host with multiple deploy keys (in your case: one for each repository). When attempting any of your designated actions (e.g., writing to repo1 or reading from repo2), all of the deploy keys listed using multiple `IdentityFile` fields will be tested sequentially, one by one, until one of them works.

## Step by step instructions

1. Create the private-public key pairs locally (on the server), using ssh-keygen. You can name the keys anything you want. Naturally, repeat this process twice (as you have two repos you which to configure access control for). Use an email address registered with your github account, for this stage (with the *-C* flag).

```bash
ssh-keygen -t ed25519 -f ~/.ssh/writeToRepo1key -C "<youremail@something>"
ssh-keygen -t ed25519 -f ~/.ssh/readFromRepo2key -C "<youremail@something>"
```

2. Make the SSH config file (if not already made)
```bash
touch ~/.ssh/config
nano ~/.ssh/config
```

3. Edit the SSH config file, adding the following host (if does not already exist), adjusting the paths to the locally-generated private key files (not .pub) accordingly.
```bash
Host github
  HostName github.com
  User git
  AddKeysToAgent yes
  PreferredAuthentications publickey
  IdentityFile ~/.ssh/writeToRepo1key
  IdentityFile ~/.ssh/readFromRepo2key
```

4. Head to `https://github.com/<organisationnameORusername>/<reponame>/settings/keys/new`, and create a new deploy key. Add a suitable title, and paste the **public key** into the "key" field. Define whether the key is granted repo *read* or repo *read-write* access through the checkbox. Do this twice, for repo1 and repo2 respectively. Then, click the **Add key** button.

```bash
cat ~/.ssh/writeToRepo1key.pub # (copy the output to add to the deploy key "key" field)
cat ~/.ssh/readFromRepo2key.pub # (copy the output to add to the deploy key "key" field)
```

5. Execute the git command related to reading / writing from the repository you wish. E.g.,
```bash
git push github:<usernameORorganization>/<repository>.git # push changes to a repository (uses *read-write* deploy key access for the repo)
git clone github:<usernameORorganization>/<repository>.git # clone a repository to local (uses *read* deploy key access for the repo)
git pull github:<usernameORorganization>/<repository>.git # pull repository changes to local (uses *read* deploy key access for the repo)
```

### Notes
Note that inside the ssh config file, we set the **User** field to "**git**". This means you don't have to write out the following when executing git commands related to a repository:
```bash
git clone git@github:<usernameORorganization>/<repository>.git 
```
And instead do this:
```bash
git clone github:<usernameORorganization>/<repository>.git 
```
Saves a bit of typing, is all.

### Citations:
- [Deploy Keys for multiple repos on 1 server](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/managing-deploy-keys#using-multiple-repositories-on-one-server)
- [My personal Unix setup cheatsheet](https://github.com/elendil7/Unix-Setup-Cheatsheet)
- [StackOverflow - multiple SSH keys on 1 client (host)](https://stackoverflow.com/questions/2419566/best-way-to-use-multiple-ssh-private-keys-on-one-client)
